### ISSUE TITLE
can't start after network-online (Fedora 31)

### ISSUE THREAD

**Issue by mikelpr** [christgau/wsdd#18]
I couldn't get the wsdd.service unit to work on startup, it would fail with the following:
```
oct 04 14:10:31 principal python3[849]: detected unhandled Python exception in '/usr/local/bin/wsdd'
oct 04 14:10:31 principal wsdd[849]: Traceback (most recent call last):
oct 04 14:10:31 principal wsdd[849]:   File "/usr/local/bin/wsdd", line 840, in <module>
oct 04 14:10:31 principal wsdd[849]:     sys.exit(main())
oct 04 14:10:31 principal wsdd[849]:   File "/usr/local/bin/wsdd", line 833, in main
oct 04 14:10:31 principal wsdd[849]:     retval = serve_wsd_requests(addresses)
oct 04 14:10:31 principal wsdd[849]:   File "/usr/local/bin/wsdd", line 776, in serve_wsd_requests
oct 04 14:10:31 principal wsdd[849]:     http_srv = klass(interface.listen_address, WSDHttpRequestHandler)
oct 04 14:10:31 principal wsdd[849]:   File "/usr/lib64/python3.7/socketserver.py", line 452, in __init__
oct 04 14:10:31 principal wsdd[849]:     self.server_bind()
oct 04 14:10:31 principal wsdd[849]:   File "/usr/local/bin/wsdd", line 64, in server_bind
oct 04 14:10:31 principal wsdd[849]:     super().server_bind()
oct 04 14:10:31 principal wsdd[849]:   File "/usr/lib64/python3.7/http/server.py", line 137, in server_bind
oct 04 14:10:31 principal wsdd[849]:     socketserver.TCPServer.server_bind(self)
oct 04 14:10:31 principal wsdd[849]:   File "/usr/lib64/python3.7/socketserver.py", line 466, in server_bind
oct 04 14:10:31 principal wsdd[849]:     self.socket.bind(self.server_address)
oct 04 14:10:31 principal wsdd[849]: OSError: [Errno 99] Cannot assign requested address
```

I went ahead and added a time.sleep(10) right after parse_args() on main() as a stop gap measure, but there's got to be a proper fix.

**Comment by christgau** [comment]
> I couldn't get the wsdd.service unit to work on startup

Do you use the unit file from the repo or a modified/other version?

> OSError: [Errno 99]

is EADDRNOTAVAIL under Linux. Wsdd enumerates the available adresses when it starts and filters out those which are not be used (the `-i`, `-4`, and `-6` options impact the filtering) and `bind`s an HTTP server for every detected IPv4/v6 address. In consequence, binding to a non-existing address _should_ not occur. In addition, the `network-online` target and dependency in the systemd unit file ...

> [...] ensure that all configured network devices are up and have an IP address assigned before  the service is started. (see [systemd documentation](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/))

I can imagine a race condition in which the systemd dependency is fulfilled but the network address changes while wsdd is starting up. Since you can "solve" the problem with a timeout, this supports the theory to some extent. You could verify this, if you add a `print` call before the http server is created (line 776) which displays the address that will be used for binding. If there is a mismatch between the printed and the available network addresses (`ip addr`) than this is the problem. 

Please also note that wsdd currently not able to deal with network changes, especially changes of the hosts IP addresses

**Comment by mikelpr** [comment]
I'm using the one from the repo, just modified it to /usr/local/bin and added a workgroup with -w

I'll check the address printing as soon as I can and get back to you

**Comment by christgau** [comment]
The same problem has been reported in #23 by @bwaindwain. To get some more details I'd like to know what interfaces are installed on the systems in question. Are there WiFi interfaces and is there IPv6 connectivity on those interfaces? It would be helpful to have `-vv` added to the `ExecStart` line of the systemd unit file. The output in the journal may help me for further analysis.

**Comment by bwaindwain** [comment]
I'm running my Ubuntu 18.04 server on a Toshiba S55 laptop. Running `ifconfig` gives `enp8s0`, `wlp7s0`, and `lo`. The server is connected by wire to my router with a reserved IP address. 

I don't know about IPv6. How do I check that? It's just a small home network with 4 computers and a bunch of other devices. 

I temporarily removed the `time.sleep(10)`, added -vv and rebooted. Attached is the log.

[wsdd.log](https://github.com/christgau/wsdd/files/4024104/wsdd.log)

**Comment by bwaindwain** [comment]
I tried adding `-i enp8s0` to `ExecStart` in wsdd.service but it still fails on reboot.

**Comment by mikelpr** [comment]
I'm sorry, I won't be able to I've moved the network to AD
keep going with bwaindwain

**Comment by bwaindwain** [comment]
In the DHCP settings of my router I've set a reserved IP address for my server. Would that affect this?

**Comment by christgau** [comment]
> I'm sorry, I won't be able to I've moved the network to AD
> keep going with bwaindwain

@mikelpr You're welcome. Thanks for your input, anyways!

**Comment by christgau** [comment]
Thanks for the log file.

> I don't know about IPv6. How do I check that? 

`ip` is the way to go here: `ip addr`. 

> It's just a small home network with 4 computers and a bunch of other devices.

If you do nothing else the IPv6 stack will configure itself using mechanisms like stateless address auto configuration (SLAAC). So even if you do nothing and your OS has IPv6 enabled, which is the case by default nowadays, you end up with IPv6 addresses on your interfaces.

> In the DHCP settings of my router I've set a reserved IP address for my server. Would that affect this?

Not necessarily. The problem would be solved if you configure the static IP on the host itself, e.g. in a  address sub-range of your network that is not handled by the DHCP server. However, it would not help much to isolate the actual problem. Thus, it would be nice, if you stick with (IPv4) DHCP for the moment, although I suspect it is IPv6 that is causing issues here. 

However, could you add the following arguments individually to the ExecStart line in the systemd unit file. Please keep the `-vv` options you already specified to produce useful output in the journal . 

* `-i enp8s0 -4` and
* `-i enp8s0 -6`

Please do the reboot for each line and report the according log files. From the log file you already provided, I would assume that the line with `-4` should work, but the one with `-6` should result in the error you already know. To be sure that things actually work (or don't) it would be helpful, if you reboot the machine several times since I don't know if the error occurs in a deterministic fashion.

**Comment by bwaindwain** [comment]
Sorry for the delay in replying. `ip addr` gives

> 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
>     link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
>     inet 127.0.0.1/8 scope host lo
>        valid_lft forever preferred_lft forever
>     inet6 ::1/128 scope host
>        valid_lft forever preferred_lft forever
> 2: enp8s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
>     link/ether c4:54:44:6c:45:bd brd ff:ff:ff:ff:ff:ff
>     inet 192.168.0.40/24 brd 192.168.0.255 scope global noprefixroute enp8s0
>        valid_lft forever preferred_lft forever
>     inet6 fe80::b66e:9f0d:2b98:4aca/64 scope link noprefixroute
>        valid_lft forever preferred_lft forever
> 3: wlp7s0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
>     link/ether 30:3a:64:07:65:89 brd ff:ff:ff:ff:ff:ff

When I add `-4` to the ExecStart I can see my Ubuntu server from my Windows machine. I rebooted three times and it works great each time.

When I switch it from `-4` to `-6` it doesn't work. I can't see my Ubuntu server from Windows. Even rebooting three times it still doesn't work. 

Here are the logs files.
[wsdd-4.log](https://github.com/christgau/wsdd/files/4062657/wsdd-4.log)
[wsdd-6.log](https://github.com/christgau/wsdd/files/4062658/wsdd-6.log)

**Comment by christgau** [comment]
Perfect! Thanks a lot for your feedback and your efforts in producing it.

> Sorry for the delay in replying. `ip addr` gives

Please, don't feel pressured. Answer whenever possible for you.

> When I add `-4` to the ExecStart I can see my Ubuntu server from my Windows machine. I rebooted three times and it works great each time.

Fine. That's what I've expected.

> When I switch it from `-4` to `-6` it doesn't work. I can't see my Ubuntu server from Windows. Even rebooting three times it still doesn't work.

The logs confirm that this is "only" an IPv6 issue. Some research confirmed that in fact a race condition is causing the problem here. When the kernel automatically assigns an IPv6 address to an interface, the IPv6 implementation checks if there are other hosts on the interface's network with the same address. This is called duplicate address detection (DAD). As long as the DAD is not completed, one cannot bind to the "preliminary" address. According requests result in error 99 on Linux. https://www.agwa.name/blog/post/beware_the_ipv6_dad_race_condition contains a nice summary of the problem.

The DAD is why you face the issue on reboot only, i.e. when the DAD is performed. Manual restarts of wsdd are much likely to happen after the DAD has completed and will therefore not face the problem.

This also explains why your workaround with a call to `sleep` before running all the network stuff in wsdd solves the issue. You can also add a line like `ExecStartPre=/bin/sleep 2` to the systemd unit file (or similar for other init systems). This may not hurt the overall experience since the host can still be discovered - even some few seconds later. However, a real solution in the code would respect such situations as well as other changes to interfaces (e.g. changes WiFi networks) but this is more a little more difficult to implement in a portable way for FreeBSD and Linux and it is worth a separate issue.

As a temporary workaround, I propose to change the unit file as hinted above. I'll add an according comment to the template unit file in `etc` and close that issue afterwards.

**Comment by palmtop** [comment]
I have had the same issue. I added a try and a loop around starting the HTTP server, but #25 will be right solution.

**Comment by christgau** [comment]
Please have a look at 6c8aaa0 (note the different branch) which should solve the issue for Linux. It currently does not work on FreeBSD but is WIP.

**Comment by christgau** [comment]
bc9a642 makes wsdd work on FreeBSD again. Although, there are some glitches (basically the same as before). Essentially, FreeBSD is still affected by the cause of this issue: When a tentative address is assigned to an interface, wsdd tries to bind to that address which fails, but there is no (elegant) way to wait for that address become available - if it does. This is, unfortunately, different to Linux. Nevertheless, the problem does not seem to appear on FreeBSD at service startup. At least of the machine available to me.

**Comment by christgau** [comment]
solved with #25
